### PR TITLE
add the rust implementation of the user CLI

### PR DESCRIPTION
**Pull request by StanPlatinum**

### CODE REVIEW

File: user-cli/rust/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+use std::num::NonZeroU32;
+
+use nix::unistd::Uid;
+
+mod backend;
+use crate::backend::device::*;
+use crate::backend::dm::*;
+
+#[macro_use]
+extern crate clap;
```
Comment (tatetian) [inline user-cli/rust/src/main.rs:12]: `#[macro_use]` is generally not needed in Rust Edition 2021.

File: user-cli/rust/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+use clap::App;
```
Comment (tatetian) [inline user-cli/rust/src/main.rs:13]: Reorder and regroup the use and module declarations in the following form.

```rust
// Import external APIs first
use data_encoding::HEXLOWER;
use ring::{digest, pbkdf2};
use std::num::NonZeroU32;
use nix::unistd::Uid;
use clap::App;

// Then, declare the modules of this crate
mod backend;

// Then, import internal APIs
use crate::backend::device::*;
use crate::backend::dm::*;
```

By the way, remember to run `cargo fmt`, which will sort use declarations automatically.

**Review by tatetian — COMMENTED** [review]
Nice done! The Rust version of the setup tool looks great. As a Rust developer who is writing his first real-world project, you have done an excellent job. I have raised some minor issues. Most of them are just tips for writing Rust code.

File: user-cli/rust/src/backend/dm.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        offset,
+        size,
+        "jindisk".into(),
+        dm_table_params,
+    )];
+
+    let name = DmName::new(dmname).expect("is valid DM name");
+    let id = DevId::Name(name);
+
+    dm.device_create(name, None, DmOptions::default())?;
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply user-cli/rust/src/backend/dm.rs:52]: Add comments to the three steps making the code more readable.

File: user-cli/rust/src/backend/device.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    let fd = file.as_raw_fd();
+    let mut cs = CalcSectors {
+        real: disk_size,
+        available: 0,
+    };
+    let cs_ptr = &mut cs as *mut CalcSectors;
+
+    println!("Querying {} ...", jindisk_interface);
+    unsafe {
+        let r = ioctl_jindisk_calc_sector(fd, cs_ptr);
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply user-cli/rust/src/backend/device.rs:103]: After implementing the `From<Errno>` for the `DeviceError` the unsafe blocks are shrunk to just one line.

File: user-cli/rust/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+
+    let r = dm_create_device(dm_name, tgt);
+    match r {
+        Ok(_) => return Ok(()),
+        Err(_) => {
+            println!("Set DM target {} failed!", dm_name);
+            // need to clean up the target and its DM table
+            dm_remove_device(dm_name).unwrap();
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply user-cli/rust/src/main.rs:68]: Thanks! Rollback logics are added to the `dm_create_device`. Therefore this removal is no longer needed. I delete it.

File: user-cli/rust/src/backend/device.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                println!("Real size: {}, available size: {}", cs.real, cs.available);
+                return Ok(data_size);
+            }
+            Err(errno) => return Err(ioctl_err_handler(errno, jindisk_interface)),
+        }
+    }
+}
+
+pub fn device_size_adjust(device: &str, device_offset: u64) -> Result<u64, DeviceError> {
+    let dev_size = device_info(device)?;
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply user-cli/rust/src/backend/device.rs:116]: Function names are rectified. Also, the name `device_ready` is replaced with `is_device_ready`.

File: user-cli/rust/src/backend/device.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    // every sector has 512 bytes
+    let blk_dev_sector_size = dev_size >> 9;
+
+    if device_offset >= blk_dev_sector_size {
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply user-cli/rust/src/backend/device.rs:120]: Thanks! I define the `Sectors` type as instructed. 
Some variables are renamed.
I also use the derive_more crate to implement some operations like Sub.

File: user-cli/rust/src/backend/dm.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    let action_flag = tgt.action_flag;
+
+    // be careful with the String
+    let mut params: String = key_string;
+    params += " ";
+    params += &iv_string;
+    params += " ";
+    params += &dev;
+    params += " ";
+    params += action_flag.to_string().as_str();
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply user-cli/rust/src/backend/dm.rs:30]: Thanks for the valuable advice!

File: user-cli/rust/src/main.rs (hunk 2 above)
Comment (StanPlatinum) [reply user-cli/rust/src/main.rs:13]: Thanks! Module declarations are reordered.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**